File: src/types/specialAbility/CeremonialItemSpecialAbility.ts
import * as DB from "tsondb/schema/dsl"
import { effect, levels, maximum, name, name_in_library } from "../_Activatable.js"
import { ap_value, ap_value_append, ap_value_l10n } from "../_ActivatableAdventurePointsValue.js"
import { aspectOptional } from "../_ActivatableNonMundane.js"
import { explicit_select_options, select_options } from "../_ActivatableSelectOptions.js"
import { skill_applications, skill_uses } from "../_ActivatableSkillApplicationsAndUses.js"
import { GeneralPrerequisites } from "../_Prerequisite.js"
import { NestedTranslationMap } from "../Locale.js"
import { Errata } from "../source/_Erratum.js"
import { src } from "../source/_PublicationRef.js"

export const CeremonialItemSpecialAbility = DB.Entity(import.meta.url, {
  name: "CeremonialItemSpecialAbility",
  namePlural: "CeremonialItemSpecialAbilities",
  type: () =>
    DB.Object({
      levels,
      select_options,
      explicit_select_options,
      skill_applications,
      skill_uses,
      maximum,
      aspect: aspectOptional(),
      prerequisites: DB.Optional({
        type: DB.IncludeIdentifier(GeneralPrerequisites),
      }),
      ap_value,
      src,
      translations: NestedTranslationMap(
        DB.Required,
        "CeremonialItemSpecialAbility",
        DB.Object({
          name,
          name_in_library,
          effect,
          ap_value_append,
          ap_value: ap_value_l10n,
          errata: DB.Optional({
            type: DB.IncludeIdentifier(Errata),
          }),
        }),
      ),
    }),
  instanceDisplayName: {},
  uniqueConstraints: [
    [
      {
        keyPath: "aspect",
      },
      {
        entityMapKeyPath: "translations",
        keyPathInEntityMap: "name_in_library",
        keyPathInEntityMapFallback: "name",
      },
    ],
  ],
})

File: Sources/OptolithDatabaseSchema/Types/specialAbility/CeremonialItemSpecialAbility.swift
import FileDB

@Model
public struct CeremonialItemSpecialAbility {
      levels,
      usage_type,
      select_options,
      skill_applications,
      skill_uses,
      maximum,
      aspect,
      prerequisites: Optional({
        type: IncludeIdentifier(GeneralPrerequisites),
      }),
      ap_value,

    /// The publications where you can find the entry.
    let src: PublicationRefs

    /// All translations for the entry, identified by IETF language tag (BCP47).
    @Relationship
    let translations: [String: Translation]

    struct Translation { // CeremonialItemSpecialAbilityTranslation
          name,
          name_in_library,
          effect,

        let errata: Errata?
    }
}
